File: 02_agent_patterns/01_agent_loop_basics/agent-loop-target/agent_loop_target/pagination.py
"""Bug 1 (off-by-one): split a sequence into fixed-size pages."""


def paginate(items, page_size):
    """Split ``items`` into consecutive pages of at most ``page_size`` elements.

    The final page holds the remainder when ``len(items)`` is not an exact
    multiple of ``page_size``. For example, 10 items with ``page_size=3`` yields
    four pages: ``[[0,1,2], [3,4,5], [6,7,8], [9]]``.

    Args:
        items: a sequence (list) of elements.
        page_size: maximum number of elements per page; must be positive.

    Returns:
        A list of pages (each a list). An empty input yields ``[]``.
    """
    if page_size <= 0:
        raise ValueError("page_size must be positive")
    pages = []
    num_pages = len(items) // page_size
    for p in range(num_pages):
        pages.append(items[p * page_size:(p + 1) * page_size])
    return pages

File: 02_agent_patterns/01_agent_loop_basics/agent-loop-target/go-example/pagination.go
package agentloop

// Paginate splits items into consecutive pages of at most pageSize elements.
//
// The final page holds the remainder when len(items) is not an exact multiple
// of pageSize. Paginate of 10 items with pageSize 3 yields four pages:
// [[0 1 2] [3 4 5] [6 7 8] [9]]. An empty input yields an empty slice.
//
// Paginate panics if pageSize is not positive.
func Paginate(items []int, pageSize int) [][]int {
	if pageSize <= 0 {
		panic("pageSize must be positive")
	}
	pages := [][]int{}
	numPages := len(items) / pageSize
	for p := 0; p < numPages; p++ {
		start := p * pageSize
		end := start + pageSize
		if end > len(items) {
			end = len(items)
		}
		pages = append(pages, items[start:end])
	}
	return pages
}
